### PR TITLE
Mock HTTP api

### PR DESCRIPTION
**Pull request by art-w**
This is a work in progress to mock the Equinix api independently from the side effectful backends. The idea is that we can call the Equinix api to get a json, add a mock test case and then improve the friendly api without having to query Equinix again.

The json can get really big, so perhaps we should store them in other files... they can also serve to discover what is currently missing in the friendly json parsing.

### CODE REVIEW

**Comment by maiste** [comment]
You are completely right! This is a really good idea :smile: Don't hesitate to assign me as a reviewer when you think it's ready for review.

**Comment by art-w** [comment]
Most of the calls should be mocked now! (*) I stole the json responses from this [better documentation of Equinix metal api](https://developer.equinix.com/catalog/metalv1) and it should give a good idea of what's missing in the friendly api.

- When creating a device, the `facility` is required? But it can guess the facility from the "metro" code? I hardcoded the facility to `"any"` just in case.
- Fix: some POST don't have a json as input, so we should send an empty body `""` rather than `"{}"`

(*) The `Ips` module is missing, but I don't think it's usable right now as there is no way of getting the unique id of an Equinix ip.

---

Some weird things related to creating new devices:
- [Operating systems are only compatible with some of the plans](https://metal.equinix.com/developers/docs/operating-systems/supported/) (?)
- Also the facilities, metro, and plans can be restricted per projects?

I'm not sure what we can do about those arbitrary restrictions... perhaps just add a lot of links to the Equinix documentation to warn the user of the possible failures? Or attempt to "validate" a `Device.config` with api calls to get the available facilities/metros/os/etc?

File: tests/mock/equinoxe_mock.ml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+let () =
+  Alcotest.run "mock"
+    [
+      ("errors", Mock_errors.tests);
+      ("auth", Mock_auth.tests);
+      ("orga", Mock_orga.tests);
+      ("devices", Mock_devices.tests);
+      ("projects", Mock_projects.tests);
+      ("users", Mock_users.tests);
```
Comment (maiste) [inline tests/mock/equinoxe_mock.ml:9]: It is an anticipation for the next PR :wink: 
```suggestion
      ("device", Mock_devices.tests);
      ("project", Mock_projects.tests);
      ("user", Mock_users.tests);
```

**Review by maiste — COMMENTED** [review]
This is the first part of the review. I want to finish the other PR to see what can be changed to make them work together and avoid pushing to much :+1:

File: tests/mock/mock.ml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    let token' = List.assoc "X-Auth-Token" headers in
+    if token' <> token then raise (Wrong_token token')
+
+  let address_length = String.length address
+
+  let starts_with ~prefix str =
+    let len = String.length prefix in
+    String.length str >= len && String.sub str 0 len = prefix
+
+  let check_url url =
```
Comment (maiste) [inline tests/mock/mock.ml:44]: Maybe `extract_path` is a more relevant name as it returns a value?

File: tests/mock/mock_orga.ml [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+      {
+        id = requested_id;
+        name = "Mock Orga 1";
+        account_id = "mock account id 1";
+        website = "";
+        maintenance_email = "admin@mock1.fr";
+        max_projects = 1;
+      }
+  in
+  let open E.Orga in
```
Comment (maiste) [inline tests/mock/mock_orga.ml:254]:
```suggestion
  let open E.Orga in
  let expected =
      {
        id = requested_id;
        name = "Mock Orga 1";
        account_id = "mock account id 1";
        website = "";
        maintenance_email = "admin@mock1.fr";
        max_projects = 1;
      }
  in
```

**Comment by maiste** [comment]
@art-w  thanks for this huge piece of work! It will avoid having regression in our code.

>  When creating a device, the `facility` is required? But it can guess the facility from the "metro" code? I hardcoded the facility to `"any"` just in case.

Not according to documentation, we can use both of them but the parameters are mutually exclusive (cf [doc](https://metal.equinix.com/developers/docs/locations/facilities/#using-facilities-in-place-of-metros)). It is the contrary. You can guess the *metro* from the *facility*: facility is a physical data center location whereas a *metro* is a group of *facilities*. If you give the facility to the data center, it will try to create the machine in this specific one :wink: 

The behaviour we want is the one from *metro* as it will try to create a machine at the wanted location in one of the physical data center. 

>   Fix: some POST don't have a json as input, so we should send an empty body `""` rather than `"{}"`

Indeed, thanks for the catch :+1: 
 
> (he `Ips` module is missing, but I don't think it's usable right now as there is no way of getting the unique id of an Equinix IP.

We can get the `Ip.id` using the `Devices.get_devices_id_ips` but that's not a problem if we don't test it. It's not a critical part.
 
> I'm not sure what we can do about those arbitrary restrictions... perhaps just add a lot of links to the Equinix documentation to warn the user of the possible failures? Or attempt to "validate" a `Device.config` with api calls to get the available facilities/metros/os/etc?

To keep the same logic, I don't think our API call should wrap more than one call per function. IMHO, we can add functions for os ([related documentation](https://metal.equinix.com/developers/docs/operating-systems/supported/#supported-operating-systems)) and metros ([related doc](https://metal.equinix.com/developers/docs/locations/metros/#api-endpoints)) to give user the possibility to check if it's possible or not. They can then link functions to ensure it is safe.

WDYT?

**Comment by art-w** [comment]
Yes, I agree! Thanks for the review :)

I'm expecting the mocks to get in the way of your refactoring. It would probably be easier if we first merge the friendly PR, then I will be able to rebase and fix the mocks to test and learn the new API.

**Comment by maiste** [comment]
Great so let's do it this way! My pleasure :smile: 

Sure, it should be finished on Monday afternoon. I ping you when it's ready :wink:

**Review by maiste — COMMENTED** [review]
The tests are working. I will just add a new constrainte over `cmdliner` as the new release has been made and it breaks the API.

File: tests/mock/mock.ml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  (module Equinoxe.Make (H) : MOCK_API)
+
+module Json_test = struct
+  type t = Ezjsonm.value
+
+  let pp h t = Format.fprintf h "%s" (Ezjsonm.value_to_string t)
+  let equal = Stdlib.( = )
+end
+
+let ezjsonm : Ezjsonm.value Alcotest.testable = (module Json_test)
```
[the remark this answers is not in the document]
Comment (art-w) [reply tests/mock/mock.ml]: Good catch, it's actually dead code with the new API!

**Comment by maiste** [comment]
As the `bin` was only here to test that the API works, we don"t need it anymore. We now have your `mock` and some examples to explain how it works.

**Review by maiste — APPROVED** [review]

**Pull request merged**